lms/djangoapps/completion/models.py

```diff
@@ -73,37 +129,103 @@ def submit_completion(self, user, course_key, block_key, completion):
+        """
+        Inserts and Updates the completion AggregateCompletion for the specified record.
+
+        Parameters:
+            * user (django.contrib.auth.models.User): The user for whom the
+              completion is being submitted.
+            * course_key (opaque_keys.edx.keys.CourseKey): The course in
+              which the submitted block is found.
+            * block_key (opaque_keys.edx.keys.UsageKey): The block that has had
+              its completion changed.
+            * aggregation_name (string): The name of the aggregated blocks.
+              This is set by the level that the aggregation
+              is occurring. Possible values are "course", "chapter", "sequential", "vertical"
+            * earned (float): The positive sum of the fractional completions of all descendant blocks up
+              to the value of possible.
+            * possible (float): The total sum of the possible completion values of all descendant
+              blocks that are visible to the user. This should be a positive integer.
+
+        Return Value:
+            (BlockCompletion, bool): A tuple comprising the created or updated
+            BlockCompletion object and a boolean value indicating whether the
+            object was newly created by this call.
+
+        Raises:
+            TypeError:
+                If the wrong type is passed for the parameters.
+
+            ValueError:
+                If the value of earned is greater than possible.
+
+            django.core.exceptions.ValidationError:
+                If earned / possible results in a number that is less than 0 or greater than 1 or any float is
+                less than zero.
+
+            RuntimeError:
+                If waffle.ENABLE_COMPLETION_AGGREGATION is not enabled.
+
+            django.db.DatabaseError:
+                If there was a problem getting, creating, or updating the
+                BlockCompletion record in the database.
+
+                This will also be a more specific error, as described here:
+                https://docs.djangoproject.com/en/1.11/ref/exceptions/#database-exceptions.
+                IntegrityError and OperationalError are relatively common
+                subclasses.
+        """
+        if not waffle.waffle().is_enabled(waffle.ENABLE_COMPLETION_AGGREGATION):
             raise RuntimeError(
-                "BlockCompletion.objects.submit_completion should not be called when the feature is disabled."
+                _("AggregateCompletionManager.submit_completion should "
+                  "not be called when the aggregation feature is disabled.")
             )
-        return obj, isnew
+        self.validate(user, course_key, block_key)
+        if earned > possible:
+            raise ValueError(_('Earned cannot be greater than the possible value.'))
+        if possible > 0.0:
+            percent = earned / possible
+        else:
+            percent = 0.0
```

Comment:
In the case where there are 0 possible completable blocks in a given aggregation, we want to give it a value of 1.0.  

Essentially, if you imagine a progress page, if a user has completed everything they can complete, they should see 100% for everything.  If they see 0% on a unit, they will think there is more work they need to do, and get confused when they can't figure out what to do.

Reply:
That makes very reasonable sense Cliff. That's some good end user perspective. Let me change that.